subprojects/code-quality/src/integTest/groovy/org/gradle/api/plugins/quality/checkstyle/CheckstylePluginIntegrationTest.groovy

```diff
@@ -55,4 +66,174 @@ class CheckstylePluginIntegrationTest extends WellBehavedPluginTest {
+    def "can use enable_external_dtd_load feature on task to override extension value for a task"() {
+        given:
+        buildFile """
+            apply plugin: 'checkstyle'
+
+            ${mavenCentralRepository()}
+
+            checkstyle {
+                enableExternalDtdLoad = false
+            }
+
+            tasks.withType(Checkstyle).configureEach {
+                enableExternalDtdLoad = true
+            }
+        """
+
+        file('src/main/java/org/sample/MyClass.java') << multilineJavaClass()
+        file('config/checkstyle/checkstyle-common.xml') << checkStyleCommonXml()
+        file('config/checkstyle/checkstyle.xml') << checkStyleMainXml()
+
+        when:
+        fails 'checkstyleMain'
+
+        then:
+        result.assertHasErrorOutput("Checkstyle rule violations were found. See the report at:")
+        result.assertHasErrorOutput("Checkstyle files with violations: 1")
+        result.assertHasErrorOutput("Checkstyle violations by severity: [error:2]")
+        file("build/reports/checkstyle/main.xml").assertContents(containsClass("org.sample.MyClass"))
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/21624")
+    @See("https://checkstyle.sourceforge.io/config_system_properties.html#Enable_External_DTD_load")
+    def "if use enable_external_dtd_load feature NOT enabled, error if feature used in rules XML"() {
+        given:
+        buildFile """
+            apply plugin: 'checkstyle'
+
+            ${mavenCentralRepository()}
+
+            checkstyle {
+                enableExternalDtdLoad = false
+            }
+        """
+
+        file('src/main/java/org/sample/MyClass.java') << multilineJavaClass()
+        file('config/checkstyle/checkstyle-common.xml') << checkStyleCommonXml()
+        file('config/checkstyle/checkstyle.xml') << checkStyleMainXml()
+
+        when:
+        fails 'checkstyleMain'
+
+        then:
+        failedDueToXmlDTDProcessingError()
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/21624")
+    @See("https://checkstyle.sourceforge.io/config_system_properties.html#Enable_External_DTD_load")
+    def "enable_external_dtd_load feature NOT enabled by default"() {
+        given:buildFile """
+            apply plugin: 'checkstyle'
+
+            ${mavenCentralRepository()}
+        """
+
+        file('src/main/java/org/sample/MyClass.java') << multilineJavaClass()
+        file('config/checkstyle/checkstyle-common.xml') << checkStyleCommonXml()
+        file('config/checkstyle/checkstyle.xml') << checkStyleMainXml()
+
+        when:
+        fails 'checkstyleMain'
+
+        then:
+        failedDueToXmlDTDProcessingError()
+    }
+
+    private failedDueToXmlDTDProcessingError() {
+        result.assertHasErrorOutput("A failure occurred while executing org.gradle.api.plugins.quality.internal.CheckstyleAction")
+        result.assertHasErrorOutput("java.lang.NullPointerException")
```

Comment:
Oh, this is a really crappy error message from Checkstyle. It seems like it's expected: https://github.com/checkstyle/checkstyle/issues/6474#issuecomment-466856949

WDYT about putting a `try {} catch(Throwable t) {}` around https://github.com/gradle/gradle/blob/8173aa2d11e6281edc1d0d40493a49052442c908/subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/internal/CheckstyleInvoker.groovy#L70-L96

IIRC, no exceptions should be thrown from that block when an expected failure occurs, so any exception caught from that block should be reported as an error with the tool or the configuration.